src/main/java/org/reaktivity/nukleus/http_cache/internal/proxy/cache/DefaultCache.java

```diff
@@ -145,22 +171,74 @@ public boolean matchCacheableRequest(
         final DefaultCacheEntry cacheEntry = cachedEntries.get(requestHash);
         return satisfiedByCache(requestHeaders) &&
                 cacheEntry != null &&
-                cacheEntry.canServeRequest(requestHeaders, authScope);
-
+                cacheEntry.canServeRequest(requestHeaders, authScope) &&
+                !cacheEntry.isValid();
     }
 
 
     public void purge(
         int requestHash)
     {
-        DefaultCacheEntry cacheEntry = this.cachedEntries.remove(requestHash);
-        entryCount.accept(-1);
+        DefaultCacheEntry cacheEntry = cachedEntries.remove(requestHash);
         if (cacheEntry != null)
         {
+            final int collectionHash = cacheEntry.getCollectionHash();
+            Int2ObjectHashMap collection = links.get(collectionHash);
+            collection.remove(requestHash);
+            if (collection.isEmpty())
+            {
+                links.remove(collectionHash);
+            }
```

Comment:
This logic can be simplified using `computeIfPresent`.
With current naming, `links.computeIfPresent(collectionHash, (h, m) -> m.remove(h) || m.isEmpty() ? null : m)`.

Reply:
Didn't understand why are you trying to delete actual `collectionHash` since we just want to remove requestHash. This is what I have came up with.

```
final int requestHashWithoutQuery = cacheEntry.requestHashWithoutQuery();
            Int2ObjectHashMap cachedEntriesByRequestHashFromWithoutQueryList = cachedEntriesByRequestHashWithoutQuery
                .computeIfPresent(requestHashWithoutQuery, (h, m) -> m.remove(requestHash) != null && m.isEmpty() ? null : m);
            if (cachedEntriesByRequestHashFromWithoutQueryList == null)
            {
                cachedEntriesByRequestHashWithoutQuery.remove(requestHashWithoutQuery);
            }
```

Reply:
Doesn't returning null from the mapping function in `computeIfPresent` already guarantee removal of the entry for key `requestHashWithoutQuery`? If so, then no need to check for `null` and remove again.

Reply:
There are two removing happening:

1. Removing  requestHash from requestHashWithoutQuery list 
2. Removing actual requestHashWithoutQuery from `cachedEntriesByRequestHashWithoutQuery ` it is empty. 

 returning null from the mapping function in computeIfPresent actually means that request hash is removed and its list is empty. Am I missing your point?

Reply:
My questions is why this...
```
Int2ObjectHashMap cachedEntriesByRequestHashFromWithoutQueryList = 
    cachedEntriesByRequestHashWithoutQuery.computeIfPresent(
        requestHashWithoutQuery, (h, m) -> m.remove(requestHash) != null && m.isEmpty() ? null : m);

if (cachedEntriesByRequestHashFromWithoutQueryList == null)
{
    cachedEntriesByRequestHashWithoutQuery.remove(requestHashWithoutQuery);
}
```
could not be simplified to this...
```
cachedEntriesByRequestHashWithoutQuery.computeIfPresent(
    requestHashWithoutQuery, (h, m) -> m.remove(requestHash) != null && m.isEmpty() ? null : m);
```

The reason they seem equivalent is because `computeIfPresent` will remove the entry from the map if the remapping function returns `null`. In the code above, the remapping function is:
```
(h, m) -> m.remove(requestHash) != null && m.isEmpty() ? null : m
```

Then `computeIfPresent` would only return `null` if either (a) the key `requestHashWithoutQuery` was not already populated initially, or (b) the remapping function returned `null` causing the `requestHashWithoutQuery` entry to be removed from the map.

In either case, the additional lines below seem redundant.
```
if (cachedEntriesByRequestHashFromWithoutQueryList == null)
{
    cachedEntriesByRequestHashWithoutQuery.remove(requestHashWithoutQuery);
}
```

Reply:
I got what you mean sorry I missed your point makes total sense.